File: examples/js/join.js
/**
 * MinterMC JavaScript Example Client
 * 
 * This script demonstrates how to connect to the MinterMC gRPC server
 * using Node.js to log in a bot.
 */

const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const path = require('path');

// Path to the shared .proto file
const PROTO_PATH = path.join(__dirname, '../../mintermc-protocol/src/main/proto/bot_service.proto');

// Load the protobuf definition
const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
    keepCase: true,
    longs: String,
    enums: String,
    defaults: true,
    oneofs: true
});

const botProto = grpc.loadPackageDefinition(packageDefinition);

// The service name is MinterBotService
const client = new botProto.MinterBotService(
    'localhost:50051',
    grpc.credentials.createInsecure()
);

function main() {
    console.log('--- MinterMC JS Client: Login Test ---');

    const loginRequest = {
        bot_name: 'JS_Bot',
        target_edition: 'JAVA_EDITION', // Protobuf enum string or index
        host: 'localhost',
        port: 25565
    };

    client.Login(loginRequest, (error, response) => {
        if (error) {
            console.error('Error during Login:', error.message);
            return;
        }

        if (response.success) {
            console.log('Login SUCCESS!');
            console.log('Session ID:', response.session_id);

            // After login, you can call other RPCs like GetStatus
            client.GetStatus({ session_id: response.session_id }, (err, status) => {
                if (err) {
                    console.error('Error getting status:', err.message);
                    return;
                }
                console.log('Current Position:', status.x, status.y, status.z);
                console.log('Health:', status.health);
            });
        } else {
            console.error('Login Failed:', response.error_message);
        }
    });
}

main();

File: examples/c/join.cpp
/**
 * MinterMC C++ Example Client
 * 
 * This example uses the gRPC C++ API. 
 * While possible in pure C, C++ is the standard for gRPC in the C-family.
 */

#include <iostream>
#include <memory>
#include <string>

#include <grpcpp/grpcpp.h>
#include "bot_service.grpc.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;
using net::mcsmash::mintermc::protocol::MinterBotService;
using net::mcsmash::mintermc::protocol::LoginRequest;
using net::mcsmash::mintermc::protocol::LoginResponse;
using net::mcsmash::mintermc::protocol::EditionType;
using net::mcsmash::mintermc::protocol::StatusRequest;
using net::mcsmash::mintermc::protocol::StatusResponse;

class MinterClient {
public:
    MinterClient(std::shared_ptr<Channel> channel)
        : stub_(MinterBotService::NewStub(channel)) {}

    void Login(const std::string& bot_name) {
        LoginRequest request;
        request.set_bot_name(bot_name);
        request.set_target_edition(EditionType::JAVA_EDITION);
        request.set_host("localhost");
        request.set_port(25565);

        LoginResponse response;
        ClientContext context;

        Status status = stub_->Login(&context, request, &response);

        if (status.ok()) {
            if (response.success()) {
                std::cout << "Login SUCCESS! Session ID: " << response.session_id() << std::endl;
                GetStatus(response.session_id());
            } else {
                std::cout << "Login Failed: " << response.error_message() << std::endl;
            }
        } else {
            std::cout << "gRPC Error: " << status.error_message() << std::endl;
        }
    }

    void GetStatus(const std::string& session_id) {
        StatusRequest request;
        request.set_session_id(session_id);

        StatusResponse response;
        ClientContext context;

        Status status = stub_->GetStatus(&context, request, &status_response);

        if (status.ok()) {
            std::cout << "Current Position: (" << response.x() << ", " << response.y() << ", " << response.z() << ")" << std::endl;
            std::cout << "Health: " << response.health() << std::endl;
        }
    }

private:
    std::unique_ptr<MinterBotService::Stub> stub_;
};

int main(int argc, char** argv) {
    std::cout << "--- MinterMC C++ Client: Login Test ---" << std::endl;
    
    MinterClient client(grpc::CreateChannel("localhost:50051", grpc::InsecureChannelCredentials()));
    client.Login("CPP_Bot");

    return 0;
}
